products/editions/enterprise-edition/b2b-suite/guides/core/crud-service.md

```diff
@@ -0,0 +1,342 @@
+# CRUD service
+
+You can download a plugin showcasing the topic [here](https://docs.enterprise.shopware.com/exampleplugins/B2bAcl.zip).
+
+## Table of contents
+*   [The Pattern](#the-pattern)
+*   [The Entity](#the-entity)
+*   [The Repository](#the-repository)
+*   [The Validation Service](#the-validation-service)
+*   [The CRUD Service](#the-crud-service)
+
+## The Pattern
+A repeating pattern used throughout the B2B-Suite are CRUD services. The B2B-Suite ships with its own entities, and therefore provides the means to **cr**eate **u**pdate and **d**elete them. Although these entities may have special requirements, there is an exclusively used naming convention and pattern used to implement all CRUD operations.
+
+The Diagram below shows the usually implemented objects with their outside dependencies.
+
+![image](/.gitbook/assets/crud-service.svg)
+
+## The Entity
+There always is an entity representing the data that has to be written. Entities are uniquely identifiable storage objects, with public properties and only a few convenience functions. An example entity looks like this:
+
+```php
+<?php
+
+namespace Shopware\B2B\Role\Framework;
+
+use Shopware\B2B\Common\CrudEntity;
+
+class RoleEntity implements CrudEntity
+{
+    /**
+     * @var int
```

Comment:
Since php 7.4.3. is the minimum php version for the B2B Suite, you can remove the data type annotations and change this to e.g. public int $id;